security_monkey/datastore_utils.py

```diff
@@ -0,0 +1,248 @@
+
+    # new item doesn't yet exist in DB
+    if not result:
+        app.logger.debug("Couldn't find item: {tech}/{account}/{region}/{item} in DB.".format(
+            tech=item.index, account=item.account, region=item.region, item=item.name
+        ))
+        return True, 'durable', result
+
+    if result.latest_revision_durable_hash != durable_hash:
+        app.logger.debug("Item: {tech}/{account}/{region}/{item} in DB has a DURABLE CHANGE.".format(
+            tech=item.index, account=item.account, region=item.region, item=item.name
+        ))
+        return True, 'durable', result
+
+    elif result.latest_revision_complete_hash != complete_hash:
+        app.logger.debug("Item: {tech}/{account}/{region}/{item} in DB has an EPHEMERAL CHANGE.".format(
+            tech=item.index, account=item.account, region=item.region, item=item.name
+        ))
+        return True, 'ephemeral', result
+
+    else:
+        app.logger.debug("Item: {tech}/{account}/{region}/{item} in DB has NO CHANGE.".format(
+            tech=item.index, account=item.account, region=item.region, item=item.name
+        ))
+        return False, None, result
+
+
+def result_from_item(item, account, technology):
+    # Construct the query to obtain the specific item from the database:
+    return datastore.Item.query.filter(Item.name == item.name, Item.region == item.region,
+                                       Item.account_id == account.id, Item.tech_id == technology.id).scalar()
+
+
+def delete_old_revisions(watcher, arns, account, technology):
+    result = Item.query.filter(
+        Item.account_id == account.id,
+        Item.tech_id == technology.id,
+        Item.arn.notin_(arns)
+    ).join((ItemRevision, Item.latest_revision_id == ItemRevision.id)) \
+        .filter(ItemRevision.active == True).all()  # noqa
+
+    for db_item in result:
+        app.logger.debug("Deleting {technology}/{account}/{name}".format(
+            technology=technology.name, account=account.name, name=db_item.name
+        ))
+
+        # Create the new revision
+        config = {"Arn": db_item.arn}
+        revision = create_revision(config, db_item)
+        db_item.revisions.append(revision)
+
+        complete_hash, durable_hash = hash_item(config, watcher.ephemeral_paths)
+
+        db_item.latest_revision_complete_hash = complete_hash
+        db_item.latest_revision_durable_hash = durable_hash
+
+        db_item.latest_revision_id = revision.id
+
+        # Add the revision:
+        datastore.db.session.add(db_item)
+        datastore.db.session.add(revision)
+        datastore.db.session.commit()
+
+        # Do it again to update the latest revision ID:
+        datastore.db.session.refresh(revision)
+        db_item.latest_revision_id = revision.id
+        datastore.db.session.add(db_item)
+
+        # Find any audit issues associated with this revision, and delete it:
+        ia = ItemAudit.query.filter(ItemAudit.item_id == db_item.id).all()
```

Comment:
@monkeysecurity So shouldn't be relevant -- as deletion of audit issues only occur when the item itself is seen as deleted.